### DISCUSSION TITLE
Ref object is not accessible inside the NotifierProvider

### DISCUSSION THREAD

**Discussion by BilalMAddam** [rrousselGit/riverpod#1964] [General]
Hello everyone,

I need your help regarding the ref's access inside the NotifierProvider.

With the StateNotifierProvider, I used to do

final authProvider = StateNotifierProvider<AuthNotifier, AsyncValue<Account>>((ref) { ... });

However, with the new Notifier syntax, it seems that the ref object is not passed.

final authProvider = NotifierProvider<AuthNotifier, AsyncValue<Account>>(() { ...});

Am I missing something?

Thank you in advance!

**Comment by rrousselGit** [comment]
The ref is available within your Notifier directly

**Comment by BilalMAddam** [comment]
Thank you @rrousselGit  for your quick reply.

Basically, I'm experimenting with Andrea's architecture proposal ```https://codewithandrea.com/articles/flutter-project-structure/```

Currently, I'm using the StateNotifierProvider to create my authProvider which depends on two other providers.

```
final authProvider = StateNotifierProvider<AuthNotifier, AsyncValue<Account>>((ref) {
  return AuthNotifier(
    authService: AuthService(
      authRepository: ref.watch(authRepositoryProvider),
      prefsRepository: ref.watch(localStorageRepositoryProvider),
    ),
  );
});
```

For refrence, here is the authRepository declaration

```
final authRepositoryProvider = Provider<AuthRepository>((ref) {
  return GetIt.I<AuthRepository>();
});

```
I'm using the get_it package in order to swipe the AuthRepository implementation seemlessly at runtime using scopes...


How can I achieve the same thing using the new Notifier syntax?

**Reply by rrousselGit** [reply]
You'd have nothing like that. Notifier/NotifierProvider doesn't work like that.

Instead place your `ref.watch`s inside `Notifier.build`

**Reply by BilalMAddam** [reply]
Putting ref.watch in the build method won't solve my problem. How would I create the AuthService object?

I can create it in the build method, but then I'd need to wrap the assigment of the authService object inside an if statement to avoid
recreating it on each build call.

It would look something like that

```
  @override
  build() {
  if (_authService == null) {
      _authService = AuthService(
              authRepository: ref.watch(authRepositoryProvider),
              prefsRepository: ref.watch(localStorageRepositoryProvider),
     );
...
```
I prefer the StateNotifierProvider implementation because the code is cleaner.

So what I'm doing is bad practice?

Is it recommended to use the NotifierProvider instead of StateNotifierProvider?

**Reply by rrousselGit** [reply]
> I can create it in the build method, but then I'd need to wrap the assigment of the authService object inside an if statement to avoid

You're free to do that, but that's not a problem.  
Your StateNotifierPRovider sample recreates the AuthService everytimes.

**Reply by BilalMAddam** [reply]
Understood. Thank you for your support!
